Kubelet segfaults when Docker is not running

I was running an experiment to verify that Taints are correctly updated when Condition changes from Unreachable to NotReady and found out that Kubelet can't start when the docker is not running.

I did an experiment on CVM and:
- made `/usr/sbin/kubelet-checker.sh` no op
- made `/usr/sbin/docker-checker.sh` no op
- `service kubelet stop`
- `service docker stop`
- wait
- reverted changes to `kubelet-checker`
- `service kubelet start`

The result is following:
```
I0404 08:23:56.527639   13582 oom_linux.go:65] attempting to set "/proc/self/oom_score_adj" to "-999"
I0404 08:23:56.527669   13582 server.go:616] cloud provider determined current node name to be e2e-test-gmarek-minion-group-7br4
I0404 08:23:56.527680   13582 server.go:802] Using root directory: /var/lib/kubelet
I0404 08:23:56.527721   13582 kubelet.go:320] cloud provider determined current node name to be e2e-test-gmarek-minion-group-7br4
I0404 08:23:56.527733   13582 kubelet.go:255] Adding manifest file: /etc/kubernetes/manifests
I0404 08:23:56.527751   13582 file.go:48] Watching path "/etc/kubernetes/manifests"
I0404 08:23:56.527759   13582 kubelet.go:265] Watching apiserver
I0404 08:23:56.527778   13582 reflector.go:187] Starting reflector *v1.Pod (0s) from k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46
I0404 08:23:56.527837   13582 reflector.go:187] Starting reflector *v1.Service (0s) from k8s.io/kubernetes/pkg/kubelet/kubelet.go:382
I0404 08:23:56.527882   13582 reflector.go:187] Starting reflector *v1.Node (0s) from k8s.io/kubernetes/pkg/kubelet/kubelet.go:390
I0404 08:23:56.527889   13582 file.go:157] Reading config file "/etc/kubernetes/manifests/e2e-image-puller.manifest"
I0404 08:23:56.529573   13582 file.go:157] Reading config file "/etc/kubernetes/manifests/kube-proxy.manifest"
I0404 08:23:56.530525   13582 reflector.go:236] Listing and watching *v1.Node from k8s.io/kubernetes/pkg/kubelet/kubelet.go:390
I0404 08:23:56.530770   13582 reflector.go:236] Listing and watching *v1.Pod from k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46
I0404 08:23:56.530932   13582 reflector.go:236] Listing and watching *v1.Service from k8s.io/kubernetes/pkg/kubelet/kubelet.go:382
I0404 08:23:56.531110   13582 config.go:293] Setting pods for source file
I0404 08:23:56.531207   13582 iptables.go:175] Could not connect to D-Bus system bus: dial unix /var/run/dbus/system_bus_socket: connect: no such file or directory
I0404 08:23:56.531249   13582 kubelet.go:494] Hairpin mode set to "promiscuous-bridge"
I0404 08:23:56.531252   13582 config.go:415] Receiving a new pod "e2e-image-puller-e2e-test-gmarek-minion-group-7br4_kube-system(0bdbf2228887dbd54a6b05f67c9bf487)"
I0404 08:23:56.531274   13582 config.go:415] Receiving a new pod "kube-proxy-e2e-test-gmarek-minion-group-7br4_kube-system(b86649617afc761351dbd891acaca743)"
I0404 08:23:56.532756   13582 iptables.go:361] running iptables -C [POSTROUTING -t nat -m comment --comment kubenet: SNAT for outbound traffic from cluster -m addrtype ! --dst-type LOCAL ! -d 10.0.0.0/8 -j MASQUERADE]
I0404 08:23:56.534117   13582 plugins.go:192] Loaded network plugin "kubenet"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x15000ec]

goroutine 1 [running]:
panic(0x2db5f80, 0xc42000e060)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/pkg/kubelet/dockershim.(*dockerService).getDockerAPIVersion(0xc4206f7b80, 0x351c3b3, 0x351c3b2, 0xc41ffa0e93)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_service.go:424 +0x7c
k8s.io/kubernetes/pkg/kubelet/dockershim.(*dockerService).checkVersionCompatibility(0xc4206f7b80, 0xc420be2d60, 0x4ea5460)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_service.go:395 +0x43
k8s.io/kubernetes/pkg/kubelet/dockershim.NewDockerService(0x4d8f0a0, 0xc420a8d0e0, 0xc4200cf300, 0x18, 0x339c09b, 0x28, 0xc4208aae60, 0xc420c40070, 0x7ffdedf30d1b, 0xe, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_service.go:173 +0x36d
k8s.io/kubernetes/pkg/kubelet.NewMainKubelet(0xc420900600, 0xc420701040, 0x335ef00, 0x18, 0xc4208ab158, 0x1)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:556 +0x6a67
k8s.io/kubernetes/cmd/kubelet/app.CreateAndInitKubelet(0xc420900600, 0xc420701040, 0x0, 0xc4208ab158, 0x1, 0x1, 0x3)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:895 +0x42
k8s.io/kubernetes/cmd/kubelet/app.RunKubelet(0xc420900600, 0xc420701040, 0x0, 0x0, 0x7ffdedf30d1b)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:811 +0x451
k8s.io/kubernetes/cmd/kubelet/app.run(0xc420900600, 0xc420701040, 0xc4208abec0, 0x456eb0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:577 +0x3b0
k8s.io/kubernetes/cmd/kubelet/app.Run(0xc420900600, 0x0, 0x22, 0xc420397290)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:285 +0x39
main.main()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/kubelet.go:48 +0x92
```
Is this expected, or kubelet should normally start and report itself as 'NotReady'?

@kubernetes/sig-node-bugs @yujuhong @wojtek-t

--- a/pkg/kubelet/dockershim/docker_service.go
+++ b/pkg/kubelet/dockershim/docker_service.go
@@ -420,6 +420,9 @@ func (ds *dockerService) getDockerAPIVersion() (*semver.Version, error) {
 	} else {
 		dv, err = ds.getDockerVersion()
 	}
+	if err != nil {
+		return nil, err
+	}
 
 	apiVersion, err := semver.Parse(dv.APIVersion)
 	if err != nil {

